library/Fission/Web/App.hs

```diff
@@ -32,6 +33,7 @@ server ::
   , MonadTime               m
   , MonadLogger             m
   , MonadDNSLink            m
+  , MonadRemoteIPFS         m
```

Comment:
🤔 How was this not here before?!

Reply:
Oooh, we had actually abstracted away from this with the app classes. I’m tempted to actually make a `App.Stats` class, or even just a method on `App.Retriever`.

Reply:
yeah i'll put a method on `App.Retriever` I like the sound of that :+1:

Reply:
actually just kidding. `App.Retriever` is all DB functions and I feel like we should have a separation in our head between "hit the DB and find something out about this appID" vs "figure out some information about the app at this CID". So maybe `App.Stats` does make more sense

Reply:
hmm and the same function is used to update a user's data root. So I either we're using a `MonadRemoteIPFS` there or we're using `App.Stats` there which doesn't seem to make that much sense.

So maybe we have a `MonadIPFSStat` or similar? But that's essentially just a subset of `MonadRemoteIPFS` :thinking:

Reply:
Actually the more i think about it, we're doing the same thing for a user's data root as well. So as you mention below, we could do App.Stats & User.Stats, but we're actually doing all of these operations on a CID before the CID is associated with the user/app.  So I think it makes most sense to think of it an IPFS operation not a user/app operation. 

It might make more sense to have something like `MonadIpfsStat` or similar. But at that point, it would essentially just be some subset of `MonadRemoteIPFS`'s functionality (and less descriptive, because "Remote IPFS" gives you more info about where it's coming from). idk, thoughts? :thinking:

Reply:
> actually just kidding. `App.Retriever` is all DB functions and I feel like we should have a separation in our head between "hit the DB and find something out about this appID" vs "figure out some information about the app at this CID". So maybe `App.Stats` does make more sense

Kind of the idea is that you don’t care if it’s hitting the DB or returning a hardcoded value, or pinging another API. It’s meant to be fully abstracted away from these details. You’re “retrieving user data”, not “accessing postgres”

Reply:
> we're actually doing all of these operations on a CID before the CID is associated with the user/app

I’m not following. Aren’t we getting the stats for what the user is responsible for in general? Not the CID, the user, the app, &c. I can imagine us having things like “certain CIDs are free because deduplication or part of our standard library”. We’re still checking the remote underneath, but the semantics we want are “check the user” not “check the CID”. We still need the remote IPFS _underneath_, but it’s not the effect that we care about — it’s a detail. Unless I’m missing something?

Reply:
ok I see what you're saying. I guess it's weird to me because we'd have `App.Stat` & `User.Stat` monads that have a method 
`getSize :: CID -> m (Either Error Bytes)` with the exact same implementation that doesn't take into account if it's an App or User.

But I guess the point is we're abstracting the IPFS details away and getting it set up in such a way that we could make it take into account some app/user details in the future

> we're actually doing all of these operations on a CID before the CID is associated with the user/app

what i meant by this is, the order of operations goes:
- user sends CID that they want to update their data root to
- we check the size of the CID in IPFS
- we update the cid & size of the user's data root in the DB (& do any other side effects like DNS)

so step 2 (which is what we're discussing) is totally agnostic to if this is a user/app & in our system, the CID isn't even associated with a user/app yet

Reply:
But is that what we’re trying to do? This is for stats on the user, not stats on arbitrary CIDs. In the future, we could easily want to call this function at other times than on data update, or do other calculation as part of that number (shared storage across multiple users, company accounts, and so on)

Reply:
Also happy to hop on a call (tomorrow) to chat about the tradeoffs here

Reply:
Thinking about this a bit more (and I need to read the code again), for the use case you mentioned above where we’re checking on write, we can probably just drop the constraint completely since it’s part of the same action.

Let’s chat more tomorrow 👍